[#58] Add intermediate representation

s-and-witch opened the pull request:
Problem: There is no way to reuse the result of diff algorithm because program print it into stdout in human readable format in-place.

Solution: divide steps of finding diff between two derivations and printing it out. Now result of `diff` is tree with result of derivation comparison, that you can print than in stdout as was before, or use in your program.

s-and-witch committed:
[#58] Add intermediate representation

Problem: There is no way to reuse result of diff alghorithm because
program print it into stdout in human readable format in-place.

Solution: divide steps of finding diff between two derivations and
printing it out. Now result of `diff` is tree with result of derivation
comparison, that you can print than in stdout as was before, or use in
your program.

SuperSandro2000 commented:
How much work would it be to now pipe the diff into ``diff -u`` or something similar to have a more powerful diff algorithm?

rvem commented on src/Diff.hs:
```diff
@@ -0,0 +1,606 @@
+    both `Map`s
+-}
+innerJoin :: Ord k => Map k a -> Map k b -> Map k (a, b)
+innerJoin = Data.Map.mergeWithKey both left right
```
AFAICS, `mergeWithKey` is unsafe, see [doc](https://hackage.haskell.org/package/containers-0.6.6/docs/Data-Map-Internal.html#v:mergeWithKey), let's use a safe analogue

s-and-witch replied:
Sure, it can be replaced with `intersectionWith`, but this usage is safe and described in the same doc, as you refer on. This code was in the earlier version of nix-diff, I just didn't change it :)
So, will we change it?

rvem replied:
I don't really insist on changing it since this code existed previously

rvem commented on src/Diff.hs:
```diff
@@ -0,0 +1,606 @@
+        -- ^ Will be Nothing, if Platform does not change
+      , builderDiff     :: Maybe (Changed Builder)
+        -- ^ Will be Nothing, if Builder does not change
+      , argumentsDiff   :: Maybe [Patience.Item Argument]
```
Is it true that the list is always non-empty when there is some diff? If so, let's use `NonEmpty` instead

s-and-witch replied:
I'm not really sure, but seems like it always non-empty.

rvem replied:
Then let's use a `NonEmpty` since `Nothing` and `Just []` basically represent the same diff

rvem commented on src/Diff.hs:
```diff
@@ -0,0 +1,606 @@
+      , argumentsDiff   :: Maybe [Patience.Item Argument]
+        -- ^ Will be Nothing, if arguments are equal
+      , sourcesDiff     :: SourcesDiff
+      , inputExtraNames :: Maybe (Changed (Set Text))
```
Cannot `inputExtraNames` and `inputsDiff` be unified, so that when one derivation has an extra input, it has the corresponding item in `inputsDiff` list?

s-and-witch replied:
No, in `inputExtraNames` there are inputs, with names that exist in one derivation and doesn't in other one. And `inputsDiff` is comparison of inputs with the same name. But I can move these fields from the `DerivationDiff`'s body, as I did for `SourcesDiff`and `EnvironmentDiff`.

rvem replied:
Hmm, Do these two situations (when input exists only in one derivation and when input exists in both) render differently?

s-and-witch replied:
Yes, inputExtraNames and inputsDiff would be rendered separately.

rvem replied:
I'm still thinking that something like `Patience.Item` can be used here to cover all cases, but let's leave it for possible future refactorings since I cannot easily come up with a way to refactor existing code

rvem commented on src/Diff.hs:
```diff
@@ -0,0 +1,606 @@
+
+    > /nix/store/${32_CHARACTER_HASH}-${NAME}.drv
+-}
+buildProductName :: FilePath -> Text
```
What is the point of having both `derivationName` and `buildProductName`? 
Isn't it correct that if `buildProductName a == buildProductName b` then `derivationName a == derivationName b` is also true

s-and-witch replied:
I don't really know why in one place is used `derivationName` and in the other one - `buildProductName`. This code was in the earlier version of nix-diff, I just didn't change it. As I can see, `buildProductName` is used in diff of sources and this name will be printed in render and the same with `derivationName` and diff of inputs, so I prefer to not change this behavior or to ask code author about it.

rvem replied:
I don't really insist on changing it since this code existed previously

rvem commented on src/Diff.hs:
```diff
@@ -0,0 +1,606 @@
+
+    OutputsDiff extraOutputs (catMaybes . catMaybes $ outputDifference)
+
+mapDiff :: (a -> b) -> Patience.Item a -> Patience.Item b
```
Could you please check if this is different from `Functor` instance for `Item`?

s-and-witch replied:
I think, we can replace it with `fmap`. This code was in the earlier version of nix-diff, I just didn't change it.

s-and-witch commented:
@SuperSandro2000 do you want to diff sources and environments in other way? In that case, you need to change `diffText` function in code or restore content of files from diff tree, because all files in the resulting tree are already diffed. You can also try to use `--character-oriented` CLI flag, as I can see, it will produce more precise diffs of files.

s-and-witch commented:
@rvem what do you think about adding Generic and Data instances to all types of tree?

rvem commented:
> what do you think about adding Generic and Data instances to all types of tree?

What are the use-cases for them? :thinking:

s-and-witch commented:
> What are the use-cases for them?

We provide Diff tree functionality as library, so it will be great to have possibility to make some orphan instances with generic / use uniplate on resulting tree. But now I understand, that user can make orphan instances of Generic/Data, if he wants.

SuperSandro2000 commented:
> @SuperSandro2000 do you want to diff sources and environments in other way? In that case, you need to change `diffText` function in code or restore content of files from diff tree, because all files in the resulting tree are already diffed. You can also try to use `--character-oriented` CLI flag, as I can see, it will produce more precise diffs of files.

I basically want to fix https://github.com/Gabriella439/nix-diff/issues/42

s-and-witch commented:
> I basically want to fix https://github.com/Gabriella439/nix-diff/issues/42

You can count offsets and print only parts that differ with calculated position. I think, it can be easily done in [renderText](https://github.com/serokell/nix-diff/blob/Player205/%2358-Add-intermediate-representation/src/Render/HumanReadable.hs#L245) function by adding new flag to the render ReaderT environment and changing algorithm of text building.

SuperSandro2000 commented:
It would be easy if I would know a thing about Haskell 🙃

s-and-witch committed:
Add golden tests for this repo

Problem: we need to check if code changes will also change the behavior

Solution: add test, that will make two derivations and check if output
of program will be different of saved one.

rvem commented on golden-tests/run-test.sh:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
```
Let's add a section about these tests to README with all prerequisites needed for running (e.g. that `nix-diff` should be in `PATH`)

rvem commented on golden-tests/old-derivation/flake.nix:
```diff
@@ -0,0 +1,22 @@
+{
+  inputs = {
+    nixpkgs.url = "github:nixos/nixpkgs/nixpkgs-unstable";
```
Let's pin a specific revision here to avoid unexpected changes of `flake.lock`

s-and-witch replied:
It already pinned in flake.lock. You can change pin by editing flake.lock or doing `nix flake update`/`nix flake lock --update-input nixpkgs`, all of these ways will cause changing of flake.lock file. 
nixpkgs input in flake will be sourced from outside only if it will not explicitly listed in inputs.

rvem replied:
Yes, one can accidentally run `nix flake lock --update-input nixpkgs` in one of derivations, but not in the other, and then tests will likely start to fail

s-and-witch replied:
He will be wrong, but this change will be shown in changes of git, so I think, that this will not cause any critical errors. And what will stop him from changing one of flake.nix?

rvem replied:
> And what will stop him from changing one of fleake.nix?

You can write something like `# DON'T CHANGE THE PINNED REVISION BELOW WITHOUT REGENERATING THE 'expected-output'` and then it's much less likely be changed on accident :sweat_smile:

rvem replied:
My point is that if some dependency implicit change can be breaking, it should be pinned explicitly

s-and-witch replied:
But it isn't implicit, because flake.lock is in the same git repo.

rvem replied:
Okay, I don't insist

s-and-witch replied:
Let's get your scenario. We wanted to add tests to the CI. Someone will
1) type command `cd ./golden-tests/new-derivation; nix flake lock --update-input nixpkgs`
2) make a PR
3) CI would report about an error.
4) In changes of commit would be clearly shown changes of one flake.lock.

rvem commented on golden-tests/expected-output:
```diff
@@ -0,0 +1,1014 @@
+- /nix/store/5kwdw9nynm8195n9dvn6zzx0rbxb2cvm-nixos-system-nixos-22.11.20221013.cfea568.drv:{out}
```
Does the upstream master `nix-diff` version produce the same output?

s-and-witch replied:
Yes, I tested on version from nixpkgs

rvem replied:
Also, does this test cover all possible diff elements, i.e. there are different inputs, outputs, environment parts?

s-and-witch replied:
> Also, does this test cover all possible diff elements, i.e. there are different inputs, outputs, environment parts?

I'm not sure about that. 

> Yes, I tested on version from nixpkgs

I start to test it one another time now and get difference. Replaced `cmp -u` with `diff` and got this output
```
38c38
< 
---
>           
44c44
< 
---
>           
46,48c46,48
< 
< 
< 
---
>           
>           
>           
58c58
< 
---
>           
...
```
So I need to understand, what is the source of difference and replace `cmp -u` with `diff`, because it helps us to get more sensible error.

rvem replied:
> I'm not sure about that.

I think it's important to have tests that cover all possible cases, so please check and update tests if needed :upside_down_face:

rvem commented on golden-tests/run-test.sh:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+export OLD_DRV="$(nix-instantiate ./old-derivation/drv.nix)"
```
A small nitpick: you probably don't need to `export` them, `old_drv="$(nix-instantiate ./old-derivation/drv.nix)"` will be just enough to make it available in the scope of the remaining script, and AFAICS, these env variables aren't used anywhere in the subsequent calls
